File: server/src/Cryptic.Core/Common/DeleteAfter.cs
namespace Cryptic.Core.Common;

public enum DeleteAfter
{
    Viewing = 0,
    OneHour = 1,
    OneDay = 2,
    OneWeek = 3,
}

public static class DeleteAfterExtensions
{
    public static TimeSpan ToTimeSpan(this DeleteAfter deleteAfter)
    {
        return deleteAfter switch
        {
            DeleteAfter.Viewing => TimeSpan.FromDays(1),
            DeleteAfter.OneHour => TimeSpan.FromHours(1),
            DeleteAfter.OneDay => TimeSpan.FromDays(1),
            DeleteAfter.OneWeek => TimeSpan.FromDays(7),
            _ => throw new InvalidOperationException("Unknown delete after!"),
        };
    }
}

File: client/src/lib/common/delete-after.ts
export default class DeleteAfter {
  static readonly Viewing = new DeleteAfter(0, "Viewing");
  static readonly OneHour = new DeleteAfter(1, "An hour");
  static readonly OneDay = new DeleteAfter(2, "A day");
  static readonly OneWeek = new DeleteAfter(3, "A week");

  private constructor(
    readonly value: number,
    readonly text: string,
  ) {}

  public static listAll(): DeleteAfter[] {
    return [DeleteAfter.Viewing, DeleteAfter.OneHour, DeleteAfter.OneDay, DeleteAfter.OneWeek];
  }

  public static fromValue(value: number) {
    const matches = DeleteAfter.listAll().filter((x) => x.value === value);

    if (matches.length === 0) {
      throw new Error(`Unknown delete after value: ${value}`);
    }

    return matches[0];
  }
}
